### PR TITLE
CAL-365 Added MGMPv2 Transformers

### PR DESCRIPTION
**Pull request by jlcsmith**
#### What does this PR do?
Adds support for MGMPv2
- MGMPv2 Input Transformer
- MGMPv2 Metacard Transformer
- MGMPv2 app
- docs
- itest

#### Who is reviewing it? 
(please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)

@brendan-hofmann 
@glenhein 

#### Choose 2 committers to review/merge the PR.
(please choose ONLY two committers from below, delete the rest)

@rzwiefel 
@beyelerb
@bdeining

#### What are the relevant tickets?

[CAL-365](https://codice.atlassian.net/browse/CAL-xxx)
#### Checklist:
- [X] Documentation Updated
- [X] Update / Add Unit Tests
- [X] Update / Add Integration Tests

### CODE REVIEW

File: catalog/mgmp/catalog-transformer-mgmp/pom.xml [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                                            <value>COVEREDRATIO</value>
+                                            <minimum>0.65</minimum>
+                                        </limit>
+                                        <limit>
+                                            <counter>COMPLEXITY</counter>
+                                            <value>COVEREDRATIO</value>
+                                            <minimum>0.7</minimum>
+                                        </limit>
+                                        <limit>
+                                            <counter>LINE</counter>
```
Comment (bdeining) [inline catalog/mgmp/catalog-transformer-mgmp/pom.xml]: line has been removed from alliance

File: catalog/mgmp/catalog-transformer-mgmp/pom.xml [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                        <configuration>
+                            <haltOnFailure>true</haltOnFailure>
+                            <rules>
+                                <rule>
+                                    <element>BUNDLE</element>
+                                    <limits>
+                                        <limit>
+                                            <counter>INSTRUCTION</counter>
+                                            <value>COVEREDRATIO</value>
+                                            <minimum>0.9</minimum>
```
Comment (bdeining) [inline catalog/mgmp/catalog-transformer-mgmp/pom.xml]: nit : precision has generally been 2 decimal places

File: catalog/mgmp/catalog-transformer-mgmp/src/test/java/org/codice/alliance/catalog/mgmp/MgmpConverterTest.java [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+    //  testMetacard1.setAttribute(MgmpConstants.ITEM_STATUS, "Work In Progress");
+  }
+
+  private void generateTestMetacard2() {
+    testMetacard2 = generateBaseTestMetacard();
+    testMetacard2.setSourceId("Alliance");
+    testMetacard2.setAttribute(Security.CLASSIFICATION, "UNCLASSIFIED");
+    testMetacard2.setAttribute(Security.METADATA_ORIGINATOR_CLASSIFICATION, "invalid");
+    testMetacard2.setAttribute(Security.RESOURCE_ORIGINATOR_CLASSIFICATION, "UNCLASSIFIED");
+    // testMetacard2.setAttribute(MgmpConstants.ITEM_STATUS, "Published");
```
Comment (bdeining) [inline catalog/mgmp/catalog-transformer-mgmp/src/test/java/org/codice/alliance/catalog/mgmp/MgmpConverterTest.java]: remove

File: catalog/mgmp/catalog-transformer-mgmp/src/test/java/org/codice/alliance/catalog/mgmp/MgmpConverterTest.java [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+    testMetacard1.setAttribute(new AttributeImpl(Core.LANGUAGE, Arrays.asList("eng", "fr", "de")));
+    testMetacard1.setAttribute(GmdConstants.RESOURCE_STATUS, "completed");
+    testMetacard1.setLocation("POLYGON (( 0 0, 10 0, 10 10, 0 10, 0 0 ))");
+    testMetacard1.setAttribute(Security.RESOURCE_CLASSIFICATION, "SECRET");
+    testMetacard1.setAttribute(Security.METADATA_CLASSIFICATION, "SECRET");
+    testMetacard1.setAttribute(
+        Security.RESOURCE_ORIGINATOR_CLASSIFICATION, "security-resource-originator-classification");
+    testMetacard1.setAttribute(
+        Security.METADATA_ORIGINATOR_CLASSIFICATION, "security-metadata-originator-classification");
+    //  testMetacard1.setAttribute(MgmpConstants.ITEM_STATUS, "Work In Progress");
```
Comment (bdeining) [inline catalog/mgmp/catalog-transformer-mgmp/src/test/java/org/codice/alliance/catalog/mgmp/MgmpConverterTest.java]: remove

File: catalog/mgmp/catalog-transformer-mgmp/src/test/java/org/codice/alliance/catalog/mgmp/MgmpTransformerTest.java [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.alliance.catalog.mgmp;
+
+import static org.hamcrest.CoreMatchers.hasItem;
+import static org.hamcrest.CoreMatchers.hasItems;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.CoreMatchers.nullValue;
+import static org.junit.Assert.assertThat;
```
Comment (bdeining) [inline catalog/mgmp/catalog-transformer-mgmp/src/test/java/org/codice/alliance/catalog/mgmp/MgmpTransformerTest.java]: should be hamcrest assertThat

File: catalog/mgmp/catalog-transformer-mgmp/src/test/java/org/codice/alliance/catalog/mgmp/MgmpUtilityTest.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.alliance.catalog.mgmp;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.Matchers.hasSize;
+import static org.junit.Assert.assertThat;
```
Comment (bdeining) [inline catalog/mgmp/catalog-transformer-mgmp/src/test/java/org/codice/alliance/catalog/mgmp/MgmpUtilityTest.java]: should be hamcrest assertThat

File: catalog/mgmp/catalog-transformer-mgmp/src/test/java/org/codice/alliance/catalog/mgmp/MgmpConverterTest.java [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.alliance.catalog.mgmp;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
```
Comment (bdeining) [inline catalog/mgmp/catalog-transformer-mgmp/src/test/java/org/codice/alliance/catalog/mgmp/MgmpConverterTest.java]: should be hamcrest assertThat

File: catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpTransformer.java [hunk 8]
```diff
@@ -0,0 +1,678 @@
...
+    }
+  }
+
+  protected void addMetacardDataQuality(MetacardImpl metacard) {
+
+    try (InputStream inputStream = getSourceInputStream()) {
+      DocumentBuilderFactory domFactory = DocumentBuilderFactory.newInstance();
+      domFactory.setNamespaceAware(false);
+      try {
+        domFactory.setFeature(
```
Comment (glenhein) [inline catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpTransformer.java]: There are some other settings that have been adopted for security reasons: https://github.com/codice/ddf/blob/master/platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java#L358

File: catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpTransformer.java [hunk 9]
```diff
@@ -0,0 +1,681 @@
...
+  }
+
+  protected void addMetacardDataQuality(MetacardImpl metacard) {
+
+    try (InputStream inputStream = getSourceInputStream()) {
+      DocumentBuilderFactory domFactory = DocumentBuilderFactory.newInstance();
+      domFactory.setNamespaceAware(false);
+      try {
+        domFactory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+        domFactory.setFeature("http://xml.org/sax/features/external-general-entitie", false);
```
Comment (glenhein) [inline catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpTransformer.java]: should be `http://xml.org/sax/features/external-general-entities`

**Comment by bdeining** [comment]
build now

**Comment by jlcsmith** [comment]
build now

File: catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpConstants.java [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.TreeMap;
+import org.codice.ddf.spatial.ogc.csw.catalog.common.GmdConstants;
+
+/**
+ * These constants are only meant to be shared between {@link MgmpConverter} and {@link
```
Comment (brendan-hofmann) [inline catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpConstants.java]: ❓ Is one of these supposed to be `{@link MgmpTransformer}`?

File: catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpTransformer.java [hunk 11]
```diff
@@ -0,0 +1,681 @@
...
+  public void addMetacardCrs(
+      final XstreamPathValueTracker pathValueTracker, MetacardImpl metacard) {
+
+    super.addMetacardCrs(pathValueTracker, metacard);
+
+    ArrayList<String> mgmpTypes = new ArrayList<>();
+    addValuesFromPathToListIfPresent(
+        pathValueTracker,
+        mgmpTypes,
+        MgmpConstants.MGMP_SPATIAL_REFERENCE_SYSTEM_NUMBER_PATH,
```
Comment (brendan-hofmann) [inline catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpTransformer.java]: ✏️ I think if you extracted the constant groups here and in some of the other methods, you could clean up a lot of these method calls. It looks like all the other parameters are the same.

File: catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpConverter.java [hunk 12]
```diff
@@ -0,0 +1,1249 @@
...
+
+  private String sourceSystemName;
+
+  private BidiMap inverseMappingList;
+
+  private static final String SECURITY_MAPPING_LIST_KEY = "securityMappingList";
+
+  private static final String SOURCE_SYSTEN_NAME_KEY = "sourceSystemName";
+
+  public void refresh(Map<String, Object> properties) {
```
Comment (brendan-hofmann) [inline catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpConverter.java]: ❓ What happens if this refreshes while transforms are being run? Should it be synchronized with the entry point?

File: catalog/mgmp/catalog-transformer-mgmp/src/test/resources/schemas/mgmp.xsd [hunk 13]
```diff
@@ -0,0 +1,751 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (brendan-hofmann) [inline catalog/mgmp/catalog-transformer-mgmp/src/test/resources/schemas/mgmp.xsd]: ❓ I notice these artifacts are under test resources. Is it worth including these in `<home>/{schema|schematron}` and maybe configuring some validators for them?

File: catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpTransformer.java [hunk 14]
```diff
@@ -0,0 +1,681 @@
...
+      }
+    }
+  }
+
+  protected void addMetacardResourceSecurityReleasability(
+      final XstreamPathValueTracker pathValueTracker, MetacardImpl metacard) {
+    addMetacardSecurityReleasability(
+        pathValueTracker,
+        metacard,
+        MgmpConstants.RESOURCE_SECURITY_RELEASABILITY_PATH,
```
[the remark this answers is not in the document]
Comment (bdeining) [reply catalog/mgmp/catalog-transformer-mgmp/src/main/java/org/codice/alliance/catalog/mgmp/MgmpTransformer.java]: To my knowledge, there are no policy decisions for resource and metadata security fields yet

**Comment by rzwiefel** [comment]
@jlcsmith What's the status on this? Should this be closed until it is prioritized somewhere?

**Comment by coyotesqrl** [comment]
@jlcsmith Tagging again to ask on the status of this PR. Should it be updated and completed so it can be merged, or should it be closed at this time?

**Pull request closed without merging**